modules/javafx.graphics/src/main/java/com/sun/javafx/css/FixedCapacitySet.java

```diff
@@ -0,0 +1,569 @@
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * Provides set implementations which have a fixed capacity. These are highly optimized
+ * and only suitable for very specific use cases where the maximum size is known in advance
+ * and when items are expected to never be removed from the sets. Fixed capacity sets will
+ * throw {@link IllegalStateException} if adding an element would exceed the maximum capacity.
+ *
+ * <p>These sets do not allow {@code null} elements, and unless otherwise specified, passing
+ * {@code null} for any argument will result in a {@link NullPointerException}.
+ *
+ * <p>Specifically, these sets are optimized for holding sets of style class names, which
+ * have the following characteristics:
+ *
+ * <ul>
+ * <li>Never {@code null}</li>
+ * <li>Sets in almost all cases contain just 1 or 2 elements, only rarely containing 3 or more</li>
+ * <li>Sets are often compared using containsAll; to avoid creating an iterator when the input is not
+ *   a {@code FixedSizeSet}, the inverse function {@link #isSuperSetOf(Collection)} is provided</li>
+ * </ul>
+ *
+ * The provided implementations are optimized for memory use, fast containsAll and
+ * fast iteration. Generally, these sets will use half the memory of an equivalent
+ * {@code HashSet} (and comparable to the immutable sets provided by {@code Set.of()})
+ * as they do not require a wrapper to hold each element.
+ *
+ * <p>These sets can only be appended, reject {@code null}s, have a fixed maximum size (which will throw
+ * an exception if exceeded), and can be frozen (made read-only without using a wrapper).
+ *
+ * <p>The fall back set implementation for large sets uses open addressing. It is
+ * only lightly optimized as the expectation is that it will see little to no use.
+ * It is still preferred over {@code HashSet} due to its low memory foot print, and
+ * faster iteration.
+ *
+ * @param <T> the element type
+ */
+public sealed abstract class FixedCapacitySet<T> extends AbstractSet<T> {
+    private static final FixedCapacitySet<?> EMPTY;
+
+    static {
+        EMPTY = new Single<>();
+
+        EMPTY.freeze();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> FixedCapacitySet<T> empty() {
+        return (FixedCapacitySet<T>) EMPTY;
+    }
+
+    /**
+     * Creates a new {@link FixedCapacitySet} with the given maximum capacity.
+     * If the capacity is exceeded, fixed capacity sets do not grow, but instead
+     * throw an {@link IllegalStateException}.
+     *
+     * @param <T> the element type
+     * @param maximumCapacity the maximum possible number of elements the set can hold, cannot be negative
+     * @return a new empty set, never {@code null}
+     */
+    public static <T> FixedCapacitySet<T> of(int maximumCapacity) {
+        return maximumCapacity == 0 ? empty()
+             : maximumCapacity == 1 ? new Single<>()
+             : maximumCapacity == 2 ? new Duo<>()
```

Comment:
I can confirm that most of the code in my application uses 1-2 styleclasses.
Since all controls have one styleclass by default (some even more, think about `TextInputControl` with `text-input` -> `TextField` with `text-field`), does it make sense to also implement a specialized `Triple` class?

We also often add 2 more styleclasses, so 3 style classes seems like a usecase that happens quite often.
I also can confirm that 4 styleclasses or even more are very rare.
I only found one: (`.label` + 3 added from us)
![image](https://github.com/openjdk/jfx/assets/66004280/2d80db3d-aca6-488e-a338-0759c82c35c8)

Reply:
Thanks for having a look, it's good to have some more use cases from other applications!  The sets are used both for the style classes on `Node`s but also the set of style classes in CSS selectors (multiple style classes in  selectors are even more rare I think than multiple style classes on `Node`s).

I think from a performance perspective, `Duo` and `Hashless` are likely very closely matched, but I can do some testing in that area.  The reason I say this is that I only got a very minor boost from including `Duo`, but kept it because it is a bit more memory efficient.

If you want, I'm also curious about how many styles are used in your application (you can see this by looking at the size of `StyleClassSet#styleClasses` after all styles are loaded).

Reply:
I see, thanks for checking that out.

I checked the application, here is the result:
After starting the app:
![image](https://github.com/openjdk/jfx/assets/66004280/623049ac-c525-4854-8f06-b9712de5d746)

After using it a little bit:
![image](https://github.com/openjdk/jfx/assets/66004280/cca4df47-829a-4d9b-b937-705b6c058172)

---
I also checked out Scenebuilder after using it a little bit:
![image](https://github.com/openjdk/jfx/assets/66004280/0f8992b5-1073-4c77-b906-356b86c61389)

Reply:
Thanks, it looks like a bit less classes are in use when compared to JFXCentral (which has about 1000), but still quite a lot.  With `BitSet` that means if style classes are referenced that happen to have a high bit set, it creates an array to hold 400 bits (using 8 longs).  When few style classes are active, the sets provided by `FixedCapacitySet` will be often smaller and (due to the other optimizations as well) faster.  8 longs is sufficient to hold 8 or 16 style name references depending on the JVM (32/64 bit, compressed pointers y/n).